asio/include/xtt/asio/context.inl

```diff
@@ -0,0 +1,340 @@
+                                                                           cookie_ctx_);
+
+                    this->run_state_machine(new_rc,
+                                            bytes_requested,
+                                            io_ptr,
+                                            gpk_lookup_callback,
+                                            assign_id_callback,
+                                            success_callback,
+                                            error_callback);
+
+                    break;
+                }
+            case return_code::WANT_PREPARSEIDCLIENTATTEST:
+                {
+                    auto suite_spec = handshake_ctx_.get_suite_spec();
+                    if (!suite_spec) {
+                        async_send_error_msg([error_callback]()
+                                             {
+                                                 error_callback(return_code::UNKNOWN_SUITE_SPEC);
+                                             });
+                        return;
+                    }
+
+                    auto cert = cert_map_.find(*suite_spec);
+                    if (cert_map_.end() == cert) {
+                        async_send_error_msg([error_callback]()
+                                             {
+                                                 error_callback(return_code::BAD_CERTIFICATE);
+                                             });
+                        return;
+                    }
+
+                    return_code new_rc = handshake_ctx_.preparse_idclientattest(bytes_requested,
+                                                                                io_ptr,
+                                                                                requested_client_id_,
+                                                                                claimed_group_id_,
+                                                                                cookie_ctx_,
+                                                                                *cert->second);
+
+                    this->run_state_machine(new_rc,
+                                            bytes_requested,
+                                            io_ptr,
+                                            gpk_lookup_callback,
+                                            assign_id_callback,
+                                            success_callback,
+                                            error_callback);
+
+                    break;
+                }
+            case return_code::WANT_VERIFYGROUPSIGNATURE:
+                gpk_lookup_callback(claimed_group_id_,
+                                   requested_client_id_,
+                                   [this, gpk_lookup_callback, assign_id_callback, success_callback, error_callback]
+                                   (std::unique_ptr<group_public_key_context> gpk_ctx)
```

Comment:
This handler should take an `error_code` param.  Otherwise, the `gpk_lookup_callback` doesn't have a way to indicate specific errors.